programmer-skills/sorting-algorithms.md

```diff
@@ -0,0 +1,442 @@
+2. While `heapsize > 1`  
+&emsp;Swap the first element (mininum or maximum value depending on your implementation) with last one,  
+&emsp; reduce the `heapsize` by one,  then heapify the heap.  
+3. The array is sorted.  
+
+####Building the heap  
+
+The items in the lowest tree level have no child nodes to compare against, thus they are considered sorted. Good for us! 
+We start from `Math.floor( array.length / 2) ` counting backwards. 
+
+![Random array to heap](http://www.stoimen.com/blog/wp-content/uploads/2012/08/7.-Random-Array-to-Heap.png)   
+*Fig 9. is taken from [here](http://www.stoimen.com/blog/2012/08/07/computer-algorithms-heap-and-heapsort-data-structure/)*   
+
+####Heapify
+
+When we are building a max heap with our heapify function, we have to compare parent nodes with its children.  
+1. If the parent node has the greatest value, then do nothing.  
+2. If one of the child nodes is greater than the parent node, then swap the greater child node with its parent node.  
+3. If both of the child nodes are greater than the parent node, then swap the biggest of the child nodes with the parent node.    
+
+Figure 10. explains heapify:
+
+![heapify explained](http://www.stoimen.com/blog/wp-content/uploads/2012/08/6.-Heapify-Part-1.png)  
+*Fig 10. is taken from [here](http://www.stoimen.com/blog/2012/08/07/computer-algorithms-heap-and-heapsort-data-structure/)*  
+	
+
+
+```javascript
+function swap(array,i,j)
+{
+	var temp = array[i];
+	array[i] = array[j];
+    array[j] = temp;
+	return array;
+}
+```
+
+```javascript
+function heapify(array,heapSize,i){
+	console.log("heapSize",heapSize)
+	var l = leftChild(i);
+	var r = rightChild(i);
+	var largest;
+	if (l < heapSize && array[l]>array[i])
+		largest = l;
+	else
+		largest = i;
+	if (r < heapSize && array[r] > array[largest])
+		largest = r;
+	if (largest != i){
+		swap(array,i,largest);		
+		return heapify(array, heapSize,largest);
+	}
+}
+```
+
+```javascript
+function buildHeap(array,heapSize){
+	for (var i= Math.floor(heapSize / 2) -1; i >=0; i--){
+		console.log("within buildheap",array,i);
+		heapify(array,heapSize,i);
+	}
+}
+``` 
+
+#### Sorting
+
+Now the elements are in a proper heap, but they are not necessarily sorted (the heap property only guarantees that every parent node is
+greater than its children's, we know nothing about the relations of elements on different branches!).
+We will collect the largest elements at the end of our array in a sorted state, and store the other elements in a heap at the front of our array.
+At start, we have no large elements at the end, and the heap at the front occupies all space. We will denote the size of the heap with heapsize,
+this is array.length at the beginning, and we will decrement it by one at each iteration step.
+
+At any given iteration, the largest element remaining in the heap is its root, physically stored as the first element in the array. The array(length-heapsize-1)
```

Comment:
I'm not sure I understand this syntax (`array(length-heapsize-1)`), can you double check it? Did you mean `array.length-heapsize-1`? Or array[array.length-heapsize-1]?